File: src/fensu/agentdocs/_helpers/identity.py
"""Resolve deterministic project skill names and installation roots."""

from __future__ import annotations

import re
import tomllib
import unicodedata
from pathlib import Path

from fensu.agentdocs.constants import SKILL_NAME_PREFIX
from fensu.agentdocs.exceptions import SkillInstallError
from fensu.agentdocs.types import SkillInstallRoot
from fensu.config.exceptions import ConfigError
from fensu.config.models import Config, ConfigSource


def normalize_skill_identity(value: str) -> str:
    """Normalize one project identity to stable lowercase ASCII kebab-case."""

    decomposed: str = unicodedata.normalize("NFKD", value)
    ascii_value: str = decomposed.encode("ascii", "ignore").decode("ascii").lower()
    normalized: str = re.sub(r"[^a-z0-9]+", "-", ascii_value).strip("-")
    if not normalized:
        raise ConfigError(f"Skill identity {value!r} does not contain an ASCII letter or digit.")
    return normalized


def find_git_root(project_root: Path) -> Path | None:
    """Return the nearest parent with a directory or file `.git` marker."""

    for candidate in (project_root.resolve(), *project_root.resolve().parents):
        marker: Path = candidate / ".git"
        if marker.is_dir() or marker.is_file():
            return candidate
    return None


def resolve_install_root(
    *, value: str | None, project_root: Path, invocation_root: Path
) -> tuple[Path, Path | None]:
    """Resolve default, named, or path installation-root semantics."""

    resolved_project: Path = project_root.resolve()
    git_root: Path | None = find_git_root(resolved_project)
    if value is None:
        return (git_root or resolved_project), git_root
    if value == SkillInstallRoot.PROJECT:
        return resolved_project, git_root
    if value == SkillInstallRoot.GIT:
        if git_root is None:
            raise SkillInstallError(
                "--install-root git requires a parent Git repository containing .git."
            )
        return git_root, git_root
    configured: Path = Path(value).expanduser()
    explicit: Path = configured if configured.is_absolute() else invocation_root / configured
    return explicit.resolve(), git_root


def resolve_skill_name(
    *, config: Config, source: ConfigSource, project_root: Path, git_root: Path | None
) -> str:
    """Resolve configured, package, directory, then Git-relative project identity."""

    if config.skills.name is not None:
        return SKILL_NAME_PREFIX + normalize_skill_identity(config.skills.name)
    pyproject_name: str | None = _nearest_project_name(project_root)
    if pyproject_name is not None:
        try:
            return SKILL_NAME_PREFIX + normalize_skill_identity(pyproject_name)
        except ConfigError:
            pass
    try:
        return SKILL_NAME_PREFIX + normalize_skill_identity(source.path.parent.name)
    except ConfigError:
        fallback: str = _git_relative_fallback(project_root=project_root, git_root=git_root)
        return SKILL_NAME_PREFIX + normalize_skill_identity(fallback)


def project_prefix(*, project_root: Path, install_root: Path) -> str:
    """Return a POSIX project prefix when installation occurs at an ancestor."""

    try:
        relative: Path = project_root.resolve().relative_to(install_root.resolve())
    except ValueError:
        return ""
    return relative.as_posix() if relative.parts else ""


def _nearest_project_name(project_root: Path) -> str | None:
    for directory in (project_root.resolve(), *project_root.resolve().parents):
        path: Path = directory / "pyproject.toml"
        if not path.is_file():
            continue
        try:
            with path.open("rb") as file:
                data: object = tomllib.load(file)
        except tomllib.TOMLDecodeError as error:
            raise ConfigError(
                f"Could not parse {path} while resolving skill identity: {error}"
            ) from error
        if not isinstance(data, dict):
            continue
        project: object = data.get("project")
        if isinstance(project, dict):
            name: object = project.get("name")
            if isinstance(name, str) and name:
                return name
    return None


def _git_relative_fallback(*, project_root: Path, git_root: Path | None) -> str:
    if git_root is None:
        return project_root.resolve().as_posix()
    relative: Path = project_root.resolve().relative_to(git_root)
    return relative.as_posix() or git_root.name

File: crates/fensu-cli/src/skills/helpers/context/identity.rs
use std::env;
use std::fs;
use std::path::{Component, Path, PathBuf};

use unicode_normalization::UnicodeNormalization;

use crate::models::Config;

const GIT_INSTALL_ROOT: &str = "git";
const HOME_PATH: &str = "~";
const PROJECT_INSTALL_ROOT: &str = "project";

pub(crate) fn find_git_root(project_root: &Path) -> Option<PathBuf> {
    project_root
        .ancestors()
        .find(|candidate| {
            let marker = candidate.join(".git");
            marker.is_dir() || marker.is_file()
        })
        .map(Path::to_path_buf)
}

pub(crate) fn resolve_install_root(
    value: Option<&str>,
    project_root: &Path,
    invocation: &Path,
    git_root: Option<&Path>,
) -> Result<PathBuf, String> {
    match value {
        None => Ok(git_root.unwrap_or(project_root).to_path_buf()),
        Some(PROJECT_INSTALL_ROOT) => Ok(project_root.to_path_buf()),
        Some(GIT_INSTALL_ROOT) => git_root.map(Path::to_path_buf).ok_or_else(|| {
            "--install-root git requires a parent Git repository containing .git.".to_owned()
        }),
        Some(path) => {
            let expanded = if path == HOME_PATH || path.starts_with("~/") {
                let home = home_dir()?;
                home.join(path.trim_start_matches('~').trim_start_matches('/'))
            } else {
                PathBuf::from(path)
            };
            Ok(normalize_absolute(if expanded.is_absolute() {
                expanded
            } else {
                invocation.join(expanded)
            }))
        }
    }
}

pub(crate) fn normalize_absolute(path: PathBuf) -> PathBuf {
    let mut result = PathBuf::new();
    for component in path.components() {
        match component {
            Component::ParentDir => {
                let _ = result.pop();
            }
            Component::CurDir => {}
            _ => result.push(component.as_os_str()),
        }
    }
    result
}

pub(crate) fn home_dir() -> Result<PathBuf, String> {
    env::var_os("HOME")
        .or_else(|| env::var_os("USERPROFILE"))
        .map(PathBuf::from)
        .ok_or_else(|| {
            "Could not resolve the home directory for global skill installation.".to_owned()
        })
}

pub(crate) fn resolve_identity(
    config: &Config,
    config_path: &Path,
    project_root: &Path,
    git_root: Option<&Path>,
) -> Result<String, String> {
    if let Some(value) = &config.skills_name {
        return Ok(format!("fensu-{}", normalize_identity(value)?));
    }
    if let Some(value) = nearest_project_name(project_root)? {
        if let Ok(normalized) = normalize_identity(&value) {
            return Ok(format!("fensu-{normalized}"));
        }
    }
    if let Some(value) = config_path
        .parent()
        .and_then(Path::file_name)
        .and_then(|item| item.to_str())
    {
        if let Ok(normalized) = normalize_identity(value) {
            return Ok(format!("fensu-{normalized}"));
        }
    }
    let fallback = git_root
        .and_then(|root| project_root.strip_prefix(root).ok())
        .filter(|path| !path.as_os_str().is_empty())
        .unwrap_or(project_root)
        .to_string_lossy();
    Ok(format!("fensu-{}", normalize_identity(&fallback)?))
}

fn normalize_identity(value: &str) -> Result<String, String> {
    let decomposed = value
        .nfkd()
        .filter(char::is_ascii)
        .collect::<String>()
        .to_lowercase();
    let mut output = String::new();
    for character in decomposed.chars() {
        if character.is_ascii_alphanumeric() {
            output.push(character);
        } else if !output.is_empty() && !output.ends_with('-') {
            output.push('-');
        }
    }
    let output = output.trim_matches('-').to_owned();
    if output.is_empty() {
        return Err(format!(
            "Skill identity {value:?} does not contain an ASCII letter or digit."
        ));
    }
    Ok(output)
}

fn nearest_project_name(project_root: &Path) -> Result<Option<String>, String> {
    for directory in project_root.ancestors() {
        let path = directory.join("pyproject.toml");
        if !path.is_file() {
            continue;
        }
        let bytes = fs::read(&path).map_err(|error| error.to_string())?;
        let value = toml::from_slice::<toml::Value>(&bytes).map_err(|error| {
            format!(
                "Could not parse {} while resolving skill identity: {error}",
                path.display()
            )
        })?;
        if let Some(name) = value
            .get("project")
            .and_then(|project| project.get("name"))
            .and_then(toml::Value::as_str)
            .filter(|name| !name.is_empty())
        {
            return Ok(Some(name.to_owned()));
        }
    }
    Ok(None)
}
